File: LC0404_Sum_of_Left_Leaves/sum_of_left_leaves.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.total = 0

    def is_leaf(self, node: TreeNode) -> bool:
        return node and (not node.left) and (not node.right)

    def rSumOfLeftLeaves(self, root: TreeNode):
        if not root:
            return

        if self.is_leaf(root.left):
            self.total += root.left.val
        else:
            self.rSumOfLeftLeaves(root.left)

        if not self.is_leaf(root.right):
            self.rSumOfLeftLeaves(root.right)

    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        self.rSumOfLeftLeaves(root)
        return self.total

File: LC0404_Sum_of_Left_Leaves/sum_of_left_leaves.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int total;
public:
    Solution()
    {
        this->total = 0;
    }

    bool isLeaf(const TreeNode* const root) const 
    {
        return root != nullptr && root->left == nullptr && root->right == nullptr;
    }

    void rSumOfLeftLeaves(const TreeNode* const root)
    {
        if (root == nullptr)
        {
            return;
        }

        if (this->isLeaf(root->left))
        {
            this->total += root->left->val; // We only want the values of the left leaves.
        }
        else
        {
            this->rSumOfLeftLeaves(root->left);
        }

        // If root's right child is a leaf, we don't need to consider it at all.
        if (!this->isLeaf(root->right))
        {
            this->rSumOfLeftLeaves(root->right);
        }
    }

    int sumOfLeftLeaves(TreeNode* root) {
        this->rSumOfLeftLeaves(root);
        return this->total;
    }
};